hibernate-core/src/main/java/org/hibernate/sql/ast/spi/AbstractSqlAstWalker.java

```diff
@@ -2632,7 +2633,11 @@ private void visitLiteral(Literal literal) {
 
 				final LiteralAsParameter<Object> jdbcParameter = new LiteralAsParameter<>( literal );
 				if ( clauseStack.getCurrent() == Clause.SELECT && dialect.requiresCastingOfParametersInSelectClause() ) {
-					castFunction().render( this, Collections.singletonList( jdbcParameter ), this );
+					castFunction().render(
+							this,
+							Collections.singletonList( jdbcParameter ),
+							(SqlAstTranslator<?>) this
```

Comment:
Ugh, that's ugly. I feel like it should not be necessary. What's the error?

Reply:
The `render` method now requires a `SqlAstTranslator` rather than `SqlAstWalker`. I can try to move down this logic into `SqlAstTranslator` to avoid the cast. The `SqlAstTranslator` is also needed in `visitSelfRenderingPredicate` and `visitSelfRenderingExpression` for calling `renderToSql`. I could move these methods down into `SqlAstTranslator` as well.

Reply:
I'm _always_ ready  to work hard to remove a typecast.

Reply:
Hmm, I could do that, but that requires a few changes again so now I am questioning if we actually need both `SqlAstWalker` and `SqlAstTranslator` i.e. why not just have one contract and thus also merge `AbstractSqlAstWalker` and `AbstractSqlAstTranslator`?